sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/kllAggregates.scala

```diff
@@ -0,0 +1,405 @@
+ *
+ * See [[https://datasketches.apache.org/docs/KLL/KLLSketch.html]] for more information.
+ *
+ * @param child
+ *   child expression against which quantile computation will occur
+ * @param mutableAggBufferOffset
+ *   offset for mutable aggregation buffer
+ * @param inputAggBufferOffset
+ *   offset for input aggregation buffer
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns the KllDoublesSketch compact binary representation.
+  """,
+  examples = """
+    Examples:
+      > SELECT LENGTH(kll_sketch_to_string_double(_FUNC_(col))) > 0 FROM VALUES (1), (2), (3), (4), (5) tab(col);
+       true
+  """,
+  group = "agg_funcs",
+  since = "4.1.0")
+// scalastyle:on line.size.limit
+case class KllSketchAggDouble(
+    override val child: Expression,
+    override val mutableAggBufferOffset: Int = 0,
+    override val inputAggBufferOffset: Int = 0)
+    extends TypedImperativeAggregate[KllBufferDouble]
+        with UnaryLike[Expression]
+        with ExpectsInputTypes {
+  def this(child: Expression) = this(child, 0, 0)
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): KllSketchAggDouble =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): KllSketchAggDouble =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+  override protected def withNewChildInternal(newInput: Expression): KllSketchAggDouble =
+    copy(child = newInput)
+
+  override def dataType: DataType = BinaryType
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      TypeCollection(
+        ByteType,
+        FloatType,
+        DoubleType,
+        IntegerType,
+        LongType,
+        ShortType
+      ))
+  override def nullable: Boolean = false
+  override def prettyName: String = "kll_sketch_agg_double"
+  override def createAggregationBuffer(): KllBufferDouble = KllBufferDouble()
+
+  /**
+   * Evaluate the input row and update the KllDoublesSketch instance with the row's value.
+   * The update function only supports a subset of Spark SQL types, and an exception will be
+   * thrown for unsupported types.
+   * Note, Null values are ignored.
+   */
+  override def update(updateBuffer: KllBufferDouble, input: InternalRow): KllBufferDouble = {
+    // Return early for null values.
+    val v = child.eval(input)
+    if (v == null) {
+      return updateBuffer
+    }
+    // Handle the different data types for sketch updates.
+    val sketch: KllDoublesSketch = updateBuffer.sketch
+    child.dataType match {
+      case ByteType =>
+        sketch.update(v.asInstanceOf[Byte].toDouble)
+      case DoubleType =>
+        sketch.update(v.asInstanceOf[Double])
+      case FloatType =>
+        sketch.update(v.asInstanceOf[Float].toDouble)
+      case IntegerType =>
+        sketch.update(v.asInstanceOf[Int].toDouble)
+      case LongType =>
```

Comment:
Converting Long to Double can cause precision loss for large integer values (greater than 2 power 53). Two Longs could convert to the same double. We can consider not supporting this conversion here.

Reply:
Sure, good point, this is done.